src/qtism/runtime/tests/AssessmentTestSession.php

```diff
@@ -2412,7 +2412,11 @@ protected function nextRouteItem($ignoreBranchings = false, $ignorePreConditions
             $numberOfBranchRules = $branchRules->count();
 
             // Branchings?
-            if ($ignoreBranchings === false && $numberOfBranchRules > 0 && $this->mustApplyBranchRules() === true) {
+            if (
+                $ignoreBranchings === false &&
+                $numberOfBranchRules > 0 &&
+                $this->mustApplyBranchRules() || $branchRules->isNonLinearNavigationModeAllowed()
```

Comment:
Just a sanity check.
Did you mean what you've written, or did you mean this instead?

```
                $ignoreBranchings === false &&
                $numberOfBranchRules > 0 &&
                ($this->mustApplyBranchRules() || $branchRules->isNonLinearNavigationModeAllowed())
```

Reply:
In either case, I strongly advise using parenthesis (even if the order of logical operations is correct) to emphasize the intended logic.

Reply:
I agree with you, let me improve that

Reply:
Oh. So, it _was_ a wrong condition after all?
I mean, what you've applied now is totally different from the original condition where, should `$branchRules->isNonLinearNavigationModeAllowed()` be `true` the rest of the condition didn't matter.

Reply:
Now the question is: "why didn't the test cases fail with the original code"?

Reply:
The test cases did not fail before, cause the test part navigation mode for those cases were always linear. I've extended the tests with nonlinear support. So now, if the branch rules is coming from the tests part (las one to be applied), we ignore the linear mode restriction.

Reply:
No-no. I mean why were they working when the condition was
```
                $ignoreBranchings === false &&
                $numberOfBranchRules > 0 &&
                $this->mustApplyBranchRules() || $branchRules->isNonLinearNavigationModeAllowed()
```
?

It's very different from the current one.

Reply:
We probably do not have proper coverage for those conditions

Reply:
That's fine. Just once again want to be 100%-sure your changes are intentional.
Quick reference of the changes between the two conditions we saw in this PR before and after my comment on this line: https://3v4l.org/G5cLN

Reply:
Yes, it is intentional, before the `||` operator outside of the parentheses would invalidate all the time all the other rules (which was passing cause of low test coverage), but it only makes sense to pass if other conditions are met too :) 

Thanks for the heads up

Reply:
OK then.
I just didn't realize that from your initial response to my comment.

All clear now.

Reply:
FYI, @ ignoreBranchings and other parameters for this very method are never passed. They were introduced on 2020, but never used. I will not change them now to avoid inflate the PR, but that is also one of the reasons that condition is not caught on tests